miniapp/src/main/java/com/rakuten/tech/mobile/miniapp/storage/util/MiniAppDatabaseEncryptionUtil.kt

```diff
@@ -0,0 +1,169 @@
+import javax.crypto.spec.SecretKeySpec
+
+private const val ENCRYPTION_ALGORITHM = "AES"
+private const val SHARED_PREFERENCE_KEY = "PASSCODE"
+private const val SHARED_PREFERENCE_NAME = "MiniAppDatabase"
+private const val CIPHER_TRANSFORMATION = "AES/CBC/PKCS5Padding"
+private const val SECRET_KEY_ALGORITHM = "PBKDF2WithHmacSHA256"
+
+/**
+ * Container for everything needed for decrypting the database.
+ *
+ * @param iv initialization vector
+ * @param salt cryptographic salt
+ * @param encryptedPasscode encrypted database key
+ */
+private data class EncryptedPasscodeDataHolder(
+    val iv: String,
+    val salt: String,
+    val encryptedPasscode: String
+)
+
+private const val PASSCODE_DECRYPTION_ERROR = "Failed to decrypt database key"
+
+/**
+ * Singleton Database Encryption Util Class
+ * Using AES/CBC/PKCS5Padding algorithm to encrypt the
+ * database passcode and the encrypted passcode will be
+ * stored to the preferences to pass the same encrypted
+ * passcode lock everytime to lock the access of the
+ * database from any third party database browser app.
+ *
+ * If needed in the future then this passcode can be taken from
+ * the user with a enter passcode UI screen to access the Database.
+ */
+object MiniAppDatabaseEncryptionUtil {
+
+    private fun getKeyFromPassword(password: String, salt: ByteArray): SecretKey {
+        val factory: SecretKeyFactory = SecretKeyFactory.getInstance(SECRET_KEY_ALGORITHM)
+        val spec: KeySpec = PBEKeySpec(password.toCharArray(), salt, 65536, 256)
+        return SecretKeySpec(
+            factory.generateSecret(spec)
+                .encoded, ENCRYPTION_ALGORITHM
+        )
+    }
+
+    private fun generateIv(): IvParameterSpec {
+        val iv = ByteArray(16)
+        SecureRandom().nextBytes(iv)
+        return IvParameterSpec(iv)
+    }
+
+    private fun generateSalt(): ByteArray {
+        val salt = ByteArray(8).apply {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+                SecureRandom.getInstanceStrong().nextBytes(this)
+            } else {
+                SecureRandom().nextBytes(this)
+            }
+        }
+        return salt
+    }
+
+    private fun getPasscodeHolder(context: Context): EncryptedPasscodeDataHolder? {
+        val prefs = context.getSharedPreferences(
+            SHARED_PREFERENCE_NAME,
+            Context.MODE_PRIVATE
+        )
+
+        val serialized = prefs.getString(SHARED_PREFERENCE_KEY, null)
+        if (serialized.isNullOrBlank()) {
+            return null
+        }
+
+        return try {
+            Gson().fromJson(serialized,
+                object : TypeToken<EncryptedPasscodeDataHolder>() {}.type
+            )
+        } catch (ex: JsonSyntaxException) {
+            null
+        }
+    }
+
+    private fun saveToPreferences(context: Context, holder: EncryptedPasscodeDataHolder) {
+        val serialized = Gson().toJson(holder)
+        val prefs = context.getSharedPreferences(
```

Comment:
I think we were suppose to use 
The `password` is encrypted/stored securely using [EncryptedSharedPreferences]
The next time that the Mini App is launched, the `password` is retrieved from EncryptedSharedPreferences
Encryption method: AES256-GCM

Reply:
Actually, In the spec it’s mentioned that the passcode needs to be encrypt with a randomly generated secret key. Which means that the passcode needed to be encrypted first. When it’s already encrypted first then I think there’s no point in storing in encrypted shared preferences. 
It’s always recommended that Passcodes should always be encrypted with a Randomly generated secret key + iv + salt(randomly generated) for higher security.  
Later iv and salt can be saved to use the same to decrypt but the Secret Key should be always generated randomly while encrypting and decrypting in the runtime using the salt.
That way you make sure of higher security.

Reply:
@gopi-karmakar I think it makes things over complicated, you could easily create a masterKey using the keystore and store the passcode in a encrypted sharedPreference using that key, the whole encryption decryption could be handled by jetpack security library.